### DISCUSSION TITLE
Group Experiment Project 1

### DISCUSSION THREAD

**Discussion by robertlong** [omigroup/omi-scripting-group#19] [General]
During our meeting on [February 2nd](https://github.com/omigroup/omi-wasm-group/discussions/18) we decided to run an experiment as a group to help explore what possible APIs we could specify for exposing simple scripting operations. People are invited to create their own project using their tech/engine of choice to implement a demo of their WASM ABI design. Alternately you can submit code that defines this WASM ABI and examples of how it might be used.

# Requirements

For the ABI requirements we settled on 1 input and 3 outputs to the WASM environment.

## Input

* **Collision event / state** This could be a boolean property or means of setting/exposing a callback function that handles collision events. You should define a means of determining if an object is colliding with another object or not. If you are creating a demo, you need a way of showcasing collision events. This could be a player colliding with a collision volume or a rigid body falling and colliding with a collision volume. Whatever method you'd like to highlight this property changing and the change in outputs.

## Output

* **Audio clip playing/paused state** This could be a boolean property or function for setting the state of an audio clip's playing/paused state. If you are creating a demo you need a way of playing back audio and responding to this playing/paused state.
* **Material color* This could be a vec3 (RGB) or vec4 (RGBA) property or a function for setting the color of a specific material.
* **Transform** This could be a set of functions, a struct, or anything else you can think of for defining the position / rotation / scale / parent / children of a node in a scene graph. You should design an ABI that showcases how you might represent and mutate transforms

# Important Dates

We'll run this experiment until we decide as a group that we have gained the knowledge we needed, however there will be check-in's / milestones at the weekly WASM group meetings.

* 1st Check-in (Tuesday Feb 8th)
* 1st Milestone (Tuesday Feb 15th)

At check-ins you can give us a quick update on how your project is going and ask a few questions if needed. During milestones we'd like to see a quick presentation (slides, a screenshare of your code editor, or a demo) showcasing your work.

**Comment by humbletim** [comment]
https://gist.github.com/AdamFrisby/d2123778c600107ae1f3e7ab0aafb260
